### PR TITLE
[Cling] improve cmd history support

### PR DESCRIPTION
**Pull request by jiangyilism**
## Changes or fixes:

1. Environment variable `CLING_HISTFILE` to specify history file path.
2. Environment variable `CLING_HISTSIZE` to specify history size. The original behavior is to not set the history size explicitly.
    That is equivalent to unlimited history size according to `textinput::History::AppendToFile()`  (history file will never be pruned so it grows indefinitely). This is probably not most users expect. @Axel-Naumann, @vgvassilev do you think it makes sense to set default cling history size to something like 500 ?   (That changes current behavior.)

## Checklist:

- [X] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #

### CODE REVIEW

**Comment by jiangyilism** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
(the build can only be triggered like this by the maintainers)

**Comment by jiangyilism** [comment]
@guitargeek The build failed in a strange way.  Did jenkins build branch `[cling_cmd_history_config](https://github.com/jiangyilism/root/tree/cling_cmd_history_config)` of this pull request?   From https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/153465/testReport/projectroot.roottest.python.distrdf/spark/roottest_python_distrdf_spark_test_all/ and https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/153465/ , It built root master branch cdd1bdf3b46 . Which does not contain my branch/commits . And the failing test [roottest_python_distrdf_spark_test_all](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/153465/testReport/projectroot.roottest.python.distrdf/spark/roottest_python_distrdf_spark_test_all/) . seems related to several distrdf changes in master branch recently.

**Comment by jiangyilism** [comment]
@guitargeek , @pcanal , @Axel-Naumann Could anyone help reviewing this patch?

**Comment by vgvassilev** [comment]
@jiangyilism, thanks for your patch! Looks reasonable approach to me. Can you add some tests for it?

**Comment by jiangyilism** [comment]
@vgvassilev Test is added. Please help resume github pipeline.

**Comment by vgvassilev** [comment]
@phsft-bot build!

File: core/textinput/src/textinput/TextInput.cpp [hunk 1]
```diff
@@ -338,6 +338,16 @@ namespace textinput {
                   [](Display *D) { return D->NotifyWindowChange(); });
   }
 
+  void TextInput::SetMaxDepth(size_t maxDepth) const
```
Comment (jalopezg-git) [inline core/textinput/src/textinput/TextInput.cpp:341]: Given that those apply to history, I would suggest renaming (same for `SetPruneLength()`), as in
```suggestion
  void TextInput::SetHistoryMaxDepth(size_t maxDepth) const
```

File: interpreter/cling/lib/UserInterface/UserInterface.cpp [hunk 2]
```diff
@@ -66,6 +67,43 @@ namespace {
 
     textinput::TextInput* operator -> () { return &m_Input; }
   };
+
+  llvm::SmallString<512> GetHistoryFilePath() {
+    if (getenv("CLING_NOHISTORY")) {
+      return {};
+    }
+
+    if (const char* HistFileEnvvar{std::getenv("CLING_HISTFILE")}) {
```
Comment (jalopezg-git) [inline interpreter/cling/lib/UserInterface/UserInterface.cpp:76]: I find the suggestion below slightly more readable (but up to you to change it) :slightly_smiling_face: 
```suggestion
    if (const char* HistFileEnvvar = std::getenv("CLING_HISTFILE")) {
```

File: interpreter/cling/lib/UserInterface/UserInterface.cpp [hunk 3]
```diff
@@ -66,6 +67,43 @@ namespace {
...
+    // ~/.cling_history
+    const char* StateHome{std::getenv("XDG_STATE_HOME")};
+
+    if (!StateHome) {
+      StateHome = "~/.local/state";
+    }
+
+    llvm::SmallString<512> FilePath;
+
+    if (!llvm::sys::fs::real_path(StateHome, FilePath, true) &&
```
Comment (jalopezg-git) [inline interpreter/cling/lib/UserInterface/UserInterface.cpp:94]: I would probably separate the path resolution, i.e. `real_path()`, from the rest of the `if` statement.

File: interpreter/cling/lib/UserInterface/UserInterface.cpp [hunk 4]
```diff
@@ -97,6 +130,17 @@ namespace cling {
                       new UITabCompletion(m_MetaProcessor->getInterpreter());
     TI->SetCompletion(Completion);
 
+    if (const char* HistSizeEnvvar{std::getenv("CLING_HISTSIZE")}) {
```
Comment (jalopezg-git) [inline interpreter/cling/lib/UserInterface/UserInterface.cpp:133]:
```suggestion
    if (const char* HistSizeEnvvar = std::getenv("CLING_HISTSIZE")) {
```

**Review by jalopezg-git — COMMENTED** [review]
Many thanks for the contribution, @jiangyilism!  :slightly_smiling_face: 

In principle, LGTM, but I will defer the approval to @vgvassilev.  Could you apply the included suggestions in the interim?

About limiting the size by default to, say `500` (default also for bash), I think that it makes sense, but that can be done in a follow-up PR.

**Comment by jiangyilism** [comment]
> Many thanks for the contribution, @jiangyilism! 🙂
> 
> In principle, LGTM, but I will defer the approval to @vgvassilev. Could you apply the included suggestions in the interim?
> 

Sure. Thanks for the feedback.

**Review by vgvassilev — APPROVED** [review]
@jiangyilism thanks for your work and sorry for the delay! Lgtm! @jalopezg-git i trust you on this one ;)

File: interpreter/cling/lib/UserInterface/UserInterface.cpp (hunk 4 above)
Comment (jiangyilism) [reply interpreter/cling/lib/UserInterface/UserInterface.cpp]: Done.

File: interpreter/cling/lib/UserInterface/UserInterface.cpp (hunk 3 above)
Comment (jiangyilism) [reply interpreter/cling/lib/UserInterface/UserInterface.cpp]: Done.

File: interpreter/cling/lib/UserInterface/UserInterface.cpp (hunk 2 above)
Comment (jiangyilism) [reply interpreter/cling/lib/UserInterface/UserInterface.cpp]: Done.

File: core/textinput/src/textinput/TextInput.cpp (hunk 1 above)
Comment (jiangyilism) [reply core/textinput/src/textinput/TextInput.cpp]: Done.

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Review by jalopezg-git — APPROVED** [review]

**Pull request merged**